### PR TITLE
bpo-29901: Improve support of path-like objects in zipapp.

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
Now general path-like objects are supported, not just pathlib.Path.

### CODE REVIEW

File: Lib/zipapp.py [hunk 1]
```diff
@@ -135,10 +133,9 @@ def create_archive(source, target=None, interpreter=None, main=None):
     with _maybe_open(target, 'wb') as fd:
         _write_file_prefix(fd, interpreter)
         with zipfile.ZipFile(fd, 'w') as z:
-            root = pathlib.Path(source)
-            for child in root.rglob('*'):
-                arcname = str(child.relative_to(root))
-                z.write(str(child), arcname)
+            for child in source.rglob('*'):
+                arcname = child.relative_to(source).as_posix()
+                z.write(child, arcname)
```
Comment (pfmoore) [inline Lib/zipapp.py:138]: I assume that ```ZipFile.write()``` now takes a general path object, and doesn't require conversion to a string? The docs for that function don't explicitly state this, unlike some other functions in that module, but I assume that's just a documentation oversight.

Also, there's a minor change in behaviour here, as the archive name is always in POSIX format now, whereas before it was in OS-specific format. But IMO that's entirely reasonable (and was an oversight on my part in the original code) so I'm happy with that.

**Review by pfmoore — APPROVED** [review]
Looks good to me - I made one comment noting a couple of points, but I don't believe they are issues.

File: Lib/zipapp.py (hunk 1 above)
Comment (serhiy-storchaka) [reply Lib/zipapp.py:138]: There is no behavior change. ZipFile itself converts the archive name to POSIX format. But I think it is better to do this explicitly. Maybe implicit conversion will be deprecated in future.

Comment (pfmoore) [reply Lib/zipapp.py:138]: Ah, OK. I agree, explicit is better here.

**Pull request merged**